### PR TITLE
(WIP) lru_cache based memoize

### PR DESCRIPTION
**Pull request by alex-hofsteede**
ref: use lru_cache in memoize and return a property

This version of memoize doesn't cache the value on the property's parent
object, so it doesn't pollute the `__dict__` namespace, meaning we don't
have to clean out `__dict__` when we go to pickle/serialize the object.

It also now returns a proper `property` object which can be used to
decorate the property's setter with `@foo.setter`. Invalidating the
cache in the setter is still up to the end user.

### CODE REVIEW

File: src/sentry/utils/cache.py [hunk 1]
```diff
@@ -8,43 +8,34 @@
...
+    namespace.
 
     >>> class A(object):
     >>>     @memoize
     >>>     def func(self):
     >>>         return 'foo'
     """
+    @property
+    @lru_cache(maxsize=10)
+    def prop(*args):
```
Comment (markstory) [inline src/sentry/utils/cache.py]: Do you need to support keyword args?

File: tests/sentry/api/serializers/test_event.py [hunk 2]
```diff
@@ -100,6 +100,9 @@ def test_message_legacy(self):
 
         # create_event automatically creates the logentry interface
         del event.data['logentry']
+        # need to re-save because the serializer forces a re-fetch from
+        # nodestore which will reanmiate the property if we don't save.
```
Comment (markstory) [inline tests/sentry/api/serializers/test_event.py]: Is 'reanmiate' supposed to be 'reanimate'?

File: src/sentry/utils/cache.py (hunk 1 above)
Comment (alex-hofsteede) [reply src/sentry/utils/cache.py]: nah, is just a getter, so actually the only possible argument is `self`, so I should just make it that.

File: tests/sentry/api/serializers/test_event.py (hunk 2 above)
Comment (alex-hofsteede) [reply tests/sentry/api/serializers/test_event.py]: yep

**Comment by alex-hofsteede** [comment]
Closing this as there is a bug in the python2/functools32 version of `lru_cache` which manifests when the hash of an object changes. Still open to ideas for a memoization decorator for properties that still allows the use of getter/setter functions, and doesn't pollute the parent object's `__dict__` requiring manual removal of keys

Bug in functools can be illustrated with the following code run in python2 vs python3
```
import sys

if sys.version_info.major == 3:
    from functools import lru_cache
else:
    from functools32 import lru_cache

@lru_cache(maxsize=1)
def foo(arg):
    return "return value"

class Hashable(object):
    def __init__(self, value):
        self.value = value

    def __hash__(self):
        return self.value

h1 = Hashable(1)
foo(h1) # add h1 to cache
h1.value = 3 # but then change its hash
h2 = Hashable(2) # evicts h1
foo(h2)
```

**Pull request closed without merging**